module/localisation/BallLocalisation/src/BallLocalisation.cpp

```diff
@@ -95,82 +102,195 @@ namespace module::localisation {
+            auto ball = std::make_unique<Ball>();
+
+            // If not accepting the ball or low confidence, then use the average team guess
+            if (!accept || (low_confidence && accept_team_guess)) {
+                ball->rBWw       = field.Hfw.inverse() * average_rBFf;
+                ball->vBw        = Eigen::Vector3d::Zero();
+                ball->confidence = 0.0;  // No confidence in other teammates' guesses
+            }
+            else {
+                // Compute the time since the last update (in seconds)
+                const auto dt =
+                    std::chrono::duration_cast<std::chrono::duration<double>>(NUClear::clock::now() - last_time_update)
+                        .count();
+                last_time_update = NUClear::clock::now();
+
+                // Time update
+                ukf.time(dt);
+
+                // Measurement update
+                ukf.measure(Eigen::Vector2d(rBWw.head<2>()),
+                            cfg.ukf.noise.measurement.position,
+                            MeasurementType::BALL_POSITION());
+
+                // Get the new state, here we are assuming ball is on the ground
+                state            = BallModel<double>::StateVec(ukf.get_state());
+                ball->rBWw       = Eigen::Vector3d(state.rBWw.x(), state.rBWw.y(), fd.ball_radius);
+                ball->vBw        = Eigen::Vector3d(state.vBw.x(), state.vBw.y(), 0);
+                ball->confidence = 1.0;  // Full confidence in our own measurements
+            }
+
+            ball->time_of_measurement = last_time_update;
+            ball->Hcw                 = balls.Hcw;
+            ball->average_rBWw        = field.Hfw.inverse() * average_rBFf;
+            if (log_level <= DEBUG) {
+                log<DEBUG>("rBWw: ", ball->rBWw.x(), ball->rBWw.y(), ball->rBWw.z());
+                log<DEBUG>("vBw: ", ball->vBw.x(), ball->vBw.y(), ball->vBw.z());
+                log<DEBUG>("average rBWw: ", ball->average_rBWw.x(), ball->average_rBWw.y(), ball->average_rBWw.z());
+                emit(graph("rBWw: ", ball->rBWw.x(), ball->rBWw.y(), ball->rBWw.z()));
+                emit(graph("vBw: ", ball->vBw.x(), ball->vBw.y(), ball->vBw.z()));
+            }
+
+            emit(ball);
+        });
+
+        // Stores ball positions received from teammates
+        on<Trigger<RoboCup>, With<Field>>().then([this](const RoboCup& robocup, const Field& field) {
+            if (!cfg.use_r2r_balls) {
+                return;
+            }
+
+            // This occurs when the ball position is an echo - ignore echos
+            if (robocup.ball.confidence == 0.0) {
+                // If the ball has no confidence, then we don't care about it
+                return;
+            }
+
+            Eigen::Vector3d rBFf = robocup.ball.position.cast<double>();
+
+            // Resize the vector of guesses if it is not large enough
+            if (team_guesses.size() < robocup.current_pose.player_id) {
```

Comment:
```suggestion
            if (team_guesses.capacity() < robocup.current_pose.player_id) {
```

I think? `size` gives the number of current elements, not themaximum before needing to be reallocated